File: test-point-compiler/src/main/java/com/zrq/test/point/compiler/TestEntryPointProcessor.java
package com.zrq.test.point.compiler;

import com.google.auto.service.AutoService;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeSpec;
import com.zrq.test.point.annotation.TestEntryPoint;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.Processor;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.element.VariableElement;
import javax.tools.Diagnostic;

/**
 * 描述：生成获取带有【TestEntryPoint】标注信息列表的Helper类
 *
 * @author zhangrq
 * createTime 2020/12/7 16:20
 */
@AutoService(Processor.class)
public class TestEntryPointProcessor extends AbstractProcessor {
    private String testModelName;

    @Override
    public synchronized void init(ProcessingEnvironment processingEnv) {
        super.init(processingEnv);
        testModelName = processingEnv.getOptions().get(Constants.TEST_MODULE_NAME);
        if (testModelName == null || testModelName.length() == 0) {
            // 未设置TEST_MODEL_NAME，报错提醒
            printMessageError("此 module【build.gradle】未设置【" + Constants.TEST_MODULE_NAME + "】参数", null);
        }
    }

    @Override
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
        Set<? extends Element> elements = roundEnv.getElementsAnnotatedWith(TestEntryPoint.class);
        if (elements != null && elements.size() > 0) {
            // 有数据，创建类
            // -编辑方法
            MethodSpec.Builder methodBuilder = MethodSpec.methodBuilder(Constants.POINT_MODULE_HELPER_METHOD_NAME)
                    .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
                    .returns(ArrayList.class)
                    .addStatement("ArrayList<" + Constants.TEST_ENTRY_POINT_INFO + "> list = new ArrayList<>()");
            for (Element element : elements) {
                int type = 0;
                String className = null;
                String methodName = null;
                if (element.getKind() == ElementKind.CLASS) {
                    // 类或者接口
                    TypeElement typeElement = (TypeElement) element;
                    className = typeElement.getQualifiedName().toString();
                    if (isSubtypeActivity(typeElement)) {
                        // Activity
                        type = 1;
                    } else if (isSubtypeFragment(typeElement)) {
                        // Fragment
                        type = 2;
                    } else if (isSubtypeSupportFragment(typeElement)) {
                        // Support Fragment
                        type = 3;
                    } else {
                        // 其它类型，不支持，报错提示
                        printMessageError("标记在类上，只支持Activity、Fragment", element);
                    }
                } else if (element.getKind() == ElementKind.METHOD) {
                    // 方法
                    // -类名
                    TypeElement typeElement = (TypeElement) element.getEnclosingElement();// 父元素
                    className = typeElement.getQualifiedName().toString();
                    // -方法名
                    ExecutableElement executableElement = (ExecutableElement) element;
                    methodName = executableElement.getSimpleName().toString();
                    // -方法参数
                    List<? extends VariableElement> parameters = executableElement.getParameters();
                    if (parameters == null || parameters.size() == 0) {
                        // 无参方法
                        Set<Modifier> modifierSet = element.getModifiers();
                        if (modifierSet.contains(Modifier.STATIC)) {
                            // 静态方法
                            type = 4;
                        } else if (isSubtypeTestListFragment(typeElement)) {
                            // 非静态方法，并且是TestListActivity的子类
                            type = 5;
                        } else {
                            // 其它类型，不支持，报错提示
                            printMessageError("标记在【方法】上，只支持【静态无参方法】或者【继承TestListFragment类的，非静态无参方法】", element);
                        }
                    } else {
                        // 有参方法，不支持，报错提示
                        printMessageError("标记在【方法】上，只支持【静态无参方法】或者【继承TestListFragment类的，非静态无参方法】", element);
                    }
                }
                if (type != 0) {
                    // 要增加
                    TestEntryPoint annotation = element.getAnnotation(TestEntryPoint.class);
                    String name = annotation.value();
                    methodBuilder.addStatement("list.add(new " + Constants.TEST_ENTRY_POINT_INFO + "($L,$S,$S,$S))", type, name, className, methodName);
                }
            }
            methodBuilder.addStatement("return list");
            // -编辑类
            TypeSpec classSpec = TypeSpec.classBuilder(Constants.POINT_MODULE_HELPER_CLASS_NAME)
                    .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
                    .addMethod(methodBuilder.build())
                    .build();
            // -编辑JavaFile
            String packageName = Constants.POINT_MODULE_HELPER_PACKAGE_PREFIX
                    + testModelName.replaceAll("[-_]", ".");// replaceAll解决名称含有[-][_]
            JavaFile javaFile = JavaFile.builder(packageName, classSpec)
                    .build();
            try {
                javaFile.writeTo(processingEnv.getFiler());
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return false;
    }

    // 此Processor支持的java版本
    @Override
    public SourceVersion getSupportedSourceVersion() {
        return SourceVersion.latestSupported();
    }

    // 支持的注解类型
    @Override
    public Set<String> getSupportedAnnotationTypes() {
        return Collections.singleton(TestEntryPoint.class.getCanonicalName());
    }

    // 是否是Activity子类型
    private boolean isSubtypeActivity(TypeElement element) {
        return processingEnv.getTypeUtils().isSubtype(element.asType(), processingEnv.getElementUtils().getTypeElement(Constants.ACTIVITY).asType());
    }

    // 是否是Fragment子类型
    private boolean isSubtypeFragment(TypeElement element) {
        return processingEnv.getTypeUtils().isSubtype(element.asType(), processingEnv.getElementUtils().getTypeElement(Constants.FRAGMENT).asType());
    }

    // 是否是 Support Fragment子类型
    private boolean isSubtypeSupportFragment(TypeElement element) {
        return processingEnv.getTypeUtils().isSubtype(element.asType(), processingEnv.getElementUtils().getTypeElement(Constants.FRAGMENT_SUPPORT_ANDROIDX).asType()) ||
                processingEnv.getTypeUtils().isSubtype(element.asType(), processingEnv.getElementUtils().getTypeElement(Constants.FRAGMENT_SUPPORT_V4).asType());
    }

    // 是否是TestListFragment子类型
    private boolean isSubtypeTestListFragment(TypeElement element) {
        return processingEnv.getTypeUtils().isSubtype(element.asType(), processingEnv.getElementUtils().getTypeElement(Constants.TEST_LIST_FRAGMENT).asType());
    }

    private void printMessageError(String message, Element element) {
        processingEnv.getMessager().printMessage(Diagnostic.Kind.ERROR, "【" + TestEntryPoint.class.getSimpleName() + "】注解" + message, element);
    }
}

File: test-point-compiler/src/main/java/com/zrq/test/point/compiler/annotationprocessor/process/TestEntryPointProcessor.kt
package com.zrq.test.point.compiler.annotationprocessor.process

import com.zrq.test.point.annotation.TestEntryPoint
import com.zrq.test.point.compiler.annotationprocessor.ktx.createJavaClass
import com.zrq.test.point.compiler.annotationprocessor.ktx.isSubtypeActivity
import com.zrq.test.point.compiler.annotationprocessor.ktx.isSubtypeFragment
import com.zrq.test.point.compiler.annotationprocessor.ktx.isSubtypeSupportFragment
import com.zrq.test.point.compiler.annotationprocessor.ktx.isSubtypeTestListFragment
import com.zrq.test.point.compiler.annotationprocessor.ktx.printMessageError
import com.zrq.test.point.compiler.common.TestEntryPointInfo
import com.zrq.test.point.compiler.common.createClassTestEntryPointHelper
import javax.annotation.processing.ProcessingEnvironment
import javax.annotation.processing.RoundEnvironment
import javax.lang.model.element.Element
import javax.lang.model.element.ElementKind
import javax.lang.model.element.ExecutableElement
import javax.lang.model.element.Modifier
import javax.lang.model.element.TypeElement

/**
 * 描述：生成获取带有【TestEntryPoint】标注信息列表的Helper类
 *
 * @author zhangrq
 * createTime 2020/12/7 16:20
 */
fun processTestEntryPoint(
    processingEnv: ProcessingEnvironment,
    testModelName: String,
    roundEnv: RoundEnvironment,
) {
    val elements = roundEnv.getElementsAnnotatedWith(TestEntryPoint::class.java)
    if (elements != null && elements.isNotEmpty()) {
        // 有数据，创建类
        createClass(processingEnv, testModelName, elements)
    }
}

private fun createClass(
    processingEnv: ProcessingEnvironment,
    testModelName: String,
    elements: Set<Element>,
) {
    val testEntryPointInfoList = mutableListOf<TestEntryPointInfo>()
    for (element in elements) {
        var type = 0
        var className: String? = null
        var methodName: String? = null
        if (element.kind == ElementKind.CLASS) {
            // 类或者接口
            val typeElement = element as TypeElement
            className = typeElement.qualifiedName.toString()
            if (processingEnv.isSubtypeActivity(typeElement)) {
                // Activity
                type = 1
            } else if (processingEnv.isSubtypeFragment(typeElement)) {
                // Fragment
                type = 2
            } else if (processingEnv.isSubtypeSupportFragment(typeElement)) {
                // Support Fragment
                type = 3
            } else {
                // 其它类型，不支持，报错提示
                processingEnv.printMessageError(
                    TestEntryPoint::class,
                    "标记在类上，只支持Activity、Fragment",
                    element
                )
            }
        } else if (element.kind == ElementKind.METHOD) {
            // 方法
            val executableElement = element as ExecutableElement
            // -方法参数
            if (executableElement.isFunParametersEmpty()) {
                // 无参方法
                // -类名
                val typeElement = element.enclosingElement as TypeElement // 父元素
                className = executableElement.getQualifiedClassName()
                // -方法名
                methodName = executableElement.simpleName.toString()
                if (executableElement.isFunStatic()) {
                    // 静态方法
                    type = 4
                } else if (processingEnv.isSubtypeTestListFragment(typeElement)) {
                    // 非静态方法，并且是TestListActivity的子类
                    type = 5
                } else {
                    // 其它类型，不支持，报错提示
                    processingEnv.printMessageError(
                        TestEntryPoint::class,
                        "标记在【方法】上，只支持【静态无参方法】或者【继承TestListFragment类的，非静态无参方法】",
                        element
                    )
                }
            } else {
                // 有参方法，不支持，报错提示
                processingEnv.printMessageError(
                    TestEntryPoint::class,
                    "标记在【方法】上，只支持【静态无参方法】或者【继承TestListFragment类的，非静态无参方法】",
                    element
                )
            }
        }
        if (type != 0) {
            // 要增加
            val annotation = element.getAnnotation(TestEntryPoint::class.java)
            val name = annotation.value
            testEntryPointInfoList.add(TestEntryPointInfo(type, name, className ?: "", methodName))
        }
    }

    processingEnv.createJavaClass(
        createClassTestEntryPointHelper(
            testModelName,
            testEntryPointInfoList
        )
    )
}

private fun ExecutableElement.isFunParametersEmpty() = parameters == null || parameters.isEmpty()

private fun ExecutableElement.isFunStatic() =
    // 修饰符里包含静态修饰符，兼容是Java类。方法注解里面有JvmStatic，兼容是Kotlin类。
    modifiers.contains(Modifier.STATIC) || !getAnnotationsByType(JvmStatic::class.java).isNullOrEmpty()

private fun ExecutableElement?.getQualifiedClassName(): String? {
    val typeElement = this?.enclosingElement as? TypeElement // 父元素

    // 兼容伴生对象的全路径名称带有".Companion"
    var qualifiedNameStr = typeElement?.qualifiedName?.toString()

    if (this != null) {
        if (!getAnnotationsByType(JvmStatic::class.java).isNullOrEmpty()
            && qualifiedNameStr?.endsWith(".Companion") == true
        ) {
            // 半生对象
            qualifiedNameStr = qualifiedNameStr.replace(".Companion", "")
        }
    }
    return qualifiedNameStr
}